Add a callback to FuncJUMP_TO_CATCH

markuspf opened the pull request:
This is another bit of the GAP dynamic library code. Again without tests
and without a guarantee that this will work in HPC-GAP

fingolfin commented on src/gapstate.h:
```diff
@@ -148,6 +148,7 @@ typedef struct GAPState {
     Obj  ErrorLVars0;       // the initial ErrorLVars value, i.e. for the lvars were the break occurred
     Obj  ErrorLVars;        // ErrorLVars as modified by DownEnv / UpEnv
     Int  ErrorLLevel;       // record where on the stack ErrorLVars is relative to the top, i.e. ErrorLVars0
+    void (*JumpToCatchFunc)(); // Error catching function for GAP shared library
```
This `JumpToCatchFunc` is a bit mysterious, and of course currently never set. And this comment doesn't really help ("shared library? what shared library?")

Perhaps "Callback invoked by FuncJUMP_TO_CATCH" would be more helpful here, and then a somewhat longer explanation in "FuncJUMP_TO_CATCH", like:  "Optionally invoke a callback just before we jump. This is not used by GAP itself, but exists for use in code linking against GAP."

markuspf commented:
Not ready.

markuspf commented:
As part of an effort with @sebasguts to get `libgap` into master, I resurrected this, and rebased it on current master.

Added a longer comment about what the added callback function is supposed to do.

I am not sure this is the correct way to handle errors for a library in the long run, but at least for now this works (and is compatible with whatever Sage wants to do). If anyone has a better idea (that ideally does not require a complete rewrite of error handling right now), I am all ears.

fingolfin commented on src/gap.c:
```diff
@@ -1103,13 +1103,20 @@ Obj FuncCALL_WITH_CATCH( Obj self, Obj func, volatile Obj args )
 
+/* TL: UInt UserHasQuit; */
+/* TL: UInt UserHasQUIT; */
+UInt SystemErrorCode;
```
These three lines look like a merge artifact -- remove?

fingolfin commented on src/gapstate.h:
```diff
@@ -123,6 +123,9 @@ typedef struct GAPState {
     Obj  BaseShellContext;
     Obj  ErrorLVars;        // ErrorLVars as modified by DownEnv / UpEnv
     Int  ErrorLLevel;       // record where on the stack ErrorLVars is relative to the top, i.e. BaseShellContext
+    void (*JumpToCatchFunc)(); // Callback that is called in FuncCALL_TO_CATCH,
```
It is called in `FuncJUMP_TO_CATCH`, not CALL_...

Would you mind renaming the callback to `JumpToCatchCallback`? I find that slightly more helpful when reading the code. But it's probably subjective, so *shrug*
